proj_eclipse/src/main/java/smartmag/models/BoxModel.java

```diff
@@ -0,0 +1,245 @@
+				.where(PRODOTTO.ID.eq(b.getProd().getId())) != null)) {
+			this.box = b;
+		} else
+			throw new IllegalArgumentException("box o prodotto non valido");
+
+	}
+
+	public BoxRecord getRecord() {
+		return record;
+	}
+
+	/**
+	 * permette di creare il record del box nel database, solo se non é giá
+	 * presente
+	 * 
+	 * @throws SQLIntegrityConstraintViolationException se il record é giá
+	 *                                                  presente nel db
+	 */
+	public void createBox() throws SQLIntegrityConstraintViolationException {
+		if (isSavedInDb()) {
+			throw new SQLIntegrityConstraintViolationException(
+					"il box#" + box.getIndirizzo() + " esiste giá");
+		}
+		BoxRecord r = DSL.newRecord(BOX);
+		r.setId(box.getIndirizzo());
+		copyBoxIntoRecord(box, r);
+		r.store();
+		this.record = r;
+	}
+
+	/**
+	 * aggiorna i dati del record e del box del modello usando quelli presenti
+	 * nel db
+	 */
+	private void refreshFromDb() {
+		this.record = fetchBoxByIndirizzo(box.getIndirizzo());
+		if (this.record != null) {
+			Box b = boxFromRecord(this.record);
+			this.box.setIndirizzo(b.getIndirizzo());
+			this.box.setQuantità(b.getQuantità());
+			this.box.setProd(b.getProd());
+		}
+
+		// TODO: event
+	}
+
+	/**
+	 * permette di cambiare il prodotto di un box solo se il prodotto giá
+	 * presente nel box é effettivamente finito e il nuovo prodotto esiste nel
+	 * db
+	 * 
+	 * @param p nuovo prodotto da assegnare al box
+	 */
+	public void cambiaProdotto(Prodotto p) {
+
+		if (box != null && box.isValid() && box.getQuantità() == 0) {
+			if (p != null && p.isValid() && (DSL.select().from(PRODOTTO)
+					.where(PRODOTTO.ID.eq(p.getId()))) != null) {
+				box.setProd(p);
+			} else
+				throw new IllegalArgumentException("prodotto non valido");
+		} else
+			throw new IllegalArgumentException(
+					"attenzione: il box deve essere valido e vuoto");
+	}
+
+	/**
+	 * cancella il record dell'ordine
+	 */
+	public void deleteRecord() {
+		if (isSavedInDb()) {
+			record.delete();
+		}
+	}
+
+	/**
+	 * incrementa la quantita di prodotto in un box
+	 * 
+	 * @param quantita numero di unita di prodotto aggiunte
+	 */
+	public void rifornisci(int quantita) {
+
+		int qi = this.box.getQuantità();
+		this.box.setQuantità(qi + quantita);
+		record.setQta(box.getQuantità());
+		record.update();
+	}
+
+	/**
+	 * diminuisce la quantita di prodotto in un box solo se la richiesta puo
+	 * essere soddisfatta
+	 * 
+	 * @param qta numero di unita di prodotto da prelevare
+	 */
+	public void preleva(int qta) {
+		if (box.getQuantità() > qta) {
+			box.setQuantità(this.box.getQuantità() - qta);
+			record.setQta(box.getQuantità());
+			record.update();
+		} else {
+			throw new IllegalArgumentException(
+					"qta richiesta maggiore di quella disponibile");
+		}
+	}
+
+	/**
+	 * imposta la quantita di prodotto alla quantita passata come parametro
+	 * 
+	 * @param qta
+	 */
+	public void setQuantita(int qta) {
```

Comment:
Questa a maggior ragione, magari anche solo privata in quanto le due precedenti bastano